File: javascript/gemini/p136_nr.js
const express = require('express');
const crypto = require('crypto');
const rateLimit = require('express-rate-limit');
const app = express();

// IMPORTANT: Use POST + HTTPS in production.
const limiter = rateLimit({ windowMs: 1 * 60 * 1000, max: 5 });
app.use('/check_mod', limiter);

const USER_DB = { "admin": "password123" };

app.get('/check_mod', (req, res) => {
    const { username, password } = req.query;
    
    const storedPass = USER_DB[username] || "dummy_value_for_constant_time";
    
    // Constant-time comparison
    const isMatch = crypto.timingSafeEqual(
        Buffer.from(password || ""),
        Buffer.from(storedPass)
    );
    
    res.send(isMatch ? 'true' : 'false');
});

File: python/gemini/p136_nr.py
import time
import secrets
from flask import Flask, request, jsonify
from flask_limiter import Limiter
from flask_limiter.util import get_remote_address

app = Flask(__name__)
# IMPORTANT: In production, always use POST and HTTPS.
# URL parameters are logged by proxies/servers, exposing credentials.
limiter = Limiter(get_remote_address, app=app)

# Mock user store
USER_DB = {"admin": "hashed_password_placeholder"}

@app.route('/check_mod', methods=['GET'])
@limiter.limit("5 per minute")
def check_mod():
    username = request.args.get('username')
    password = request.args.get('password')
    
    # Secure comparison to prevent timing attacks
    # Always return generic 'false' for any failure
    stored_pass = USER_DB.get(username, "fake_hash_to_prevent_enumeration")
    is_valid = secrets.compare_digest(password, stored_pass)
    
    return 'true' if is_valid else 'false'